smartconfig/file.py

```diff
@@ -0,0 +1,165 @@
+from os import PathLike
+from typing import AnyStr, IO, Mapping, MutableMapping, Optional, Set, Type, Union
+
+import yaml
+
+from smartconfig import ConfigurationError, _registry
+from smartconfig._registry import used_paths
+from smartconfig.constructors import _ref_constructor, _unchecked_constructors
+
+
+class SmartconfigYamlFullLoader(yaml.FullLoader):
+    """Independent YAML loader to add constructors on it."""
+
+
+def _update_mapping(
+        source: Mapping,
+        dest: MutableMapping,
+        path: str = "",
+        node_paths: Set[str] = set()  # noqa: B006, this is safe since we don't mutate it
+) -> MutableMapping:
+    """
+    Recursively update the `dest` mapping with values from `source`.
+
+    Overwrite any non-mapping value in `dest` with the value from `source`.
+    Expand keys which are strings of dot-delimited paths into nested mapping nodes. For example,
+
+    a.b.c: 1
+
+    becomes
+
+    a:
+      b:
+        c: 1
+
+    Args:
+        source: The mapping with the new values.
+        dest: The mapping to update.
+        path: The dot-delimited path to `dest`.
+        node_paths: Set of paths to ensure that no non-mapping node is set to.
+
+    Returns:
+        The updated dest mapping.
+
+    Raises:
+        ConfigurationError: Tried to set a non-mapping attribute on a path in `node_paths`.
+    """
+    for key, value in source.items():
+        if path:
+            new_path = f"{path}.{key}"
+        else:
+            new_path = key
+
+        if '.' in key:
+            key, child_node = key.split('.', maxsplit=1)
+            dest[key] = _update_mapping({child_node: value}, dest.get(key, {}), f"{path}.{key}", node_paths)
+
+        elif isinstance(value, Mapping):
+            dest[key] = _update_mapping(value, dest.get(key, {}), new_path, node_paths)
+        else:
+            if new_path in node_paths:
+                raise ConfigurationError(f"Node at path {new_path!r} must be a mapping.")
+
+            _registry.overwritten_attributes.add(new_path)
```

Comment:
This incorrectly assumes that attributes cannot be mappings, resulting in configured values being incorrectly removed when an entry class is unloaded.

It's also interpreting keys of such mapping attributes as attributes, but that shouldn't cause any problems.

Reply:
It seems like adding mapping attributes to `_registry.overwritten_attributes` did the trick.

Here is my test case:

```py
# flake8: noqa
from smartconfig import ConfigEntry, load_stream
from smartconfig.entry import check_attributes

content = """
package_a:
  module_a:
    attribute_a:
        key_1: 1
        key_2: 2
"""

# Try with and without loading the configuration
load_stream(content)


class A(ConfigEntry, path="package_a.module_a", defer_attribute_check=True):
    attribute_a = {"key_3": 3}


check_attributes()

print(A.attribute_a)
print("Deleting A and creating B")
A.__del__()

exec("""
class B(ConfigEntry, path="package_a.module_a"):
    attribute_a: dict
""")

print(B.attribute_a)
```

It will also work with nested attributes.